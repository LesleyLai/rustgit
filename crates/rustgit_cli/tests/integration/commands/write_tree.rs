use crate::common::{git, populate_folder, rustgit};
use assert_cmd::prelude::*;
use std::str::from_utf8;
use test_utils::{test_path, touch};

#[test]
fn write_tree() -> anyhow::Result<()> {
    let working_dir = test_path!();
    let git = || git(&working_dir);

    git().init();

    populate_folder(&working_dir);

    git().stage(".");

    let tree_hash = rustgit(&working_dir).write_tree()?;

    let command = git().args(["ls-tree", &tree_hash]).assert().success();
    let output = from_utf8(&command.get_output().stdout)?;
    insta::assert_snapshot!(output);

    Ok(())
}

// Git actually sorts the file list for the entire project before building the tree, rather than
// sorting entries within trees themselves
#[test]
fn sort_in_correct_order() -> anyhow::Result<()> {
    let working_dir = test_path!();

    let git = || git(&working_dir);
    git().init();

    std::fs::create_dir(working_dir.join("foo"))?;
    touch(&working_dir.join("foo.txt"))?;
    touch(&working_dir.join("foo").join("bar.txt"))?;

    git().stage(".");

    let tree_hash = rustgit(&working_dir).write_tree()?;

    let command = git().args(["ls-tree", &tree_hash]).assert().success();
    let output = from_utf8(&command.get_output().stdout)?;
    insta::assert_snapshot!(output);

    Ok(())
}
