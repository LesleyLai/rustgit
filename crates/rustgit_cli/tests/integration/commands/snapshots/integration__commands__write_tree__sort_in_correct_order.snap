---
source: crates/rustgit_cli/tests/integration/commands/write_tree.rs
assertion_line: 54
expression: output
---
100644 blob e69de29bb2d1d6434b8b29ae775ad8c2e48c5391	foo.txt
040000 tree 91c9988a7a7250917d0bdea6b89e07a5cf0b3a81	foo
